WIP: Add ConvertTo-Clixml and ConvertFrom-Clixml cmdlets

charlieschmidt opened the pull request:
## PR Summary
ConvertTo-Clixml and ConvertFrom-Clixml cmdlets (to go with Export/Import-Clixml; like Import/Export/ConvertFrom/ConvertTo-Csv)

Refactored Export-Clixml into base functionality shared with ConvertTo-Clixml.

Import/ConvertFrom-Clixml both use existing ImportXmlHelper class - extended that to create MemoryStream for ConvertFrom-

Corresponding pester tests.

Closes #3898
## PR Checklist

Note: Please mark anything not applicable to this PR `NA`.

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - [x] Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] User facing [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - [X] Issue filed - Issue link: https://github.com/PowerShell/PowerShell-Docs/pull/2022
- [x] [Change is not breaking](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)
- [x] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
    - [x] [Add `[feature]` if the change is significant or affectes feature tests](https://github.com/PowerShell/PowerShell/blob/master/docs/testing-guidelines/testing-guidelines.md#requesting-additional-tests-for-a-pr)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` to the beginning of the title and remove the prefix when the PR is ready.

charlieschmidt committed:
ConvertTo-Clixml and ConvertFrom-Clixml cmdlets (to go with Export/Import-Clixml; like Import/Export/ConvertFrom/ConvertTo-Csv)
Refactored Export-Clixml into base functionality shared with ConvertTo-Clixml.
Import/ConvertFrom-Clixml both use existing ImportXmlHelper class - extended that to create MemoryStream for ConvertFrom-
Corresponding pester tests.

Closes #3898

adityapatwardhan commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/XmlCommands.cs:
```diff
@@ -808,7 +1010,25 @@ internal void Stop()
+            s.Position = 0;
+            return s;
+        }
+
```
Remove extra lines.

adityapatwardhan commented on test/powershell/Modules/Microsoft.PowerShell.Utility/clixml.tests.ps1:
```diff
@@ -123,6 +123,58 @@
+        It "converts an object to string as expected" {
+            $objClixml = ConvertTo-Clixml -Depth 1 -InputObject ($gpsList | Select-Object -First 1)
+
+            foreach($item in $objClixml)
```
Can we use `Should Match` instead of looping and `Select-String`?

adityapatwardhan commented on test/powershell/Modules/Microsoft.PowerShell.Utility/clixml.tests.ps1:
```diff
@@ -123,6 +123,58 @@
+                }
+            }
+
+            $isExisted | Should Be $true
```
Typically the error message on console when a failure happens with assertions like this are not helpful. 
`Was false when true was expected`.

adityapatwardhan commented on test/powershell/Modules/Microsoft.PowerShell.Utility/clixml.tests.ps1:
```diff
@@ -123,6 +123,58 @@
+            $isExisted | Should Be $true
+        }
+
+
```
Extra lines

adityapatwardhan commented on test/powershell/Modules/Microsoft.PowerShell.Utility/clixml.tests.ps1:
```diff
@@ -123,6 +123,58 @@
+                foreach($gpsItem in $gpsList)
+                {
+                    $checkId = $gpsItem.Id
+                    if (($null -ne $(Select-String -InputObject $item -SimpleMatch $checkId)) -and ($null -ne $(Select-String -InputObject $item -SimpleMatch "Id")))
```
Same comment as above with looping and select-string

adityapatwardhan commented on test/powershell/Modules/Microsoft.PowerShell.Utility/clixml.tests.ps1:
```diff
@@ -188,6 +240,45 @@
+            $filePath | Should Exist
+
+            $fileContent = Get-Content $filePath -raw
+            $fileContent | Should Not Be $null
```
Use `Should Not BeNullOrEmpty` instead

adityapatwardhan commented on test/powershell/Modules/Microsoft.PowerShell.Utility/clixml.tests.ps1:
```diff
@@ -188,6 +240,45 @@
+            $filePath | Should Exist
+
+            $fileContent = Get-Content $filePath
+            $fileContent | Should Not Be $null
```
Same as above.

charlieschmidt commented:
Should address the review issues, and a few other spelling/correction stuff.

I've also added documentation and a PR for PowerShell-Docs at https://github.com/PowerShell/PowerShell-Docs/pull/2022

charlieschmidt committed:
Fix comments, spelling
Change to auto properties where appropriate
Don't allow empty string as input to ConvertFrom-Clixml
Fix IDisposable implementation in ConvertFrom-Clixml
SImplify logic of ConvertFrom-Clixml
Simplier/corrected tests

adityapatwardhan commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/XmlCommands.cs:
```diff
@@ -351,10 +428,106 @@ protected override void ProcessRecord()
+            _inputObjectBuffer.Add(InputObject);
+        }
+
+
```
Extra line.

adityapatwardhan commented:
@charlieschmidt  Thanks for adding the documentation, much appreciated.

charlieschmidt commented:
@adityapatwardhan yup. Anything else needed?

charlieschmidt commented on test/powershell/Modules/Microsoft.PowerShell.Utility/clixml.tests.ps1:
```diff
@@ -123,6 +123,30 @@
+            Remove-Item $filePath -Force -ErrorAction SilentlyContinue
+        }
+
+        It "converts an object to string as expected" {
```
I don't think so, not without making it uglier.  Because the difference in the two tests isn't a variable or parameter, it's function param vs pipeline.   That said, can move some of the duplicate code to the BeforeAll, should make it more clear what is really being tested

charlieschmidt committed:
Use named parameter for boolean; remove empty lines; clean up ConvertFrom-Clixml test cases

iSazonov commented on test/powershell/Modules/Microsoft.PowerShell.Utility/clixml.tests.ps1:
```diff
@@ -158,7 +182,7 @@
 
             $fileContent = Get-Content $filePath
-            $fileContent | Should Not Be $null
+            $fileContent | Should Not BeNullOrEmpty
```
It seems it is unrelated the PR - please revert.

charlieschmidt replied:
can, but was cleaning up other instances of this, want me to pull it to a separate PR or?

https://github.com/PowerShell/PowerShell/pull/5815#discussion_r160291972

iSazonov replied:
Such changes increase the reviewing time because a reviewer must understand that this change is not related to code changes.

Thanks for clarify. I don't see the request. I am fine to leave the change.

iSazonov commented on test/powershell/Modules/Microsoft.PowerShell.Utility/clixml.tests.ps1:
```diff
@@ -172,7 +196,7 @@
 
             $fileContent = Get-Content $filePath
-            $fileContent | Should Not Be $null
+            $fileContent | Should Not BeNullOrEmpty
```
It seems it is unrelated the PR - please revert.

iSazonov replied:
Closed.

iSazonov commented on test/powershell/Modules/Microsoft.PowerShell.Utility/clixml.tests.ps1:
```diff
@@ -123,6 +123,30 @@
+            Remove-Item $filePath -Force -ErrorAction SilentlyContinue
+        }
+
+        It "converts an object to string as expected" {
```
Below we use "deserializes" - so I believe here we should use "serializes".

iSazonov commented on test/powershell/Modules/Microsoft.PowerShell.Utility/clixml.tests.ps1:
```diff
@@ -123,6 +123,30 @@
+            $objClixml | Should Match "<I32 N=`"Id`">$($gps.Id)</I32>"
+        }
+
+        It "converts an object from the pipeline to string as expected" {
```
The same - "serializes".

iSazonov commented on test/powershell/Modules/Microsoft.PowerShell.Utility/clixml.tests.ps1:
```diff
@@ -218,6 +266,22 @@ Describe "Deserializing corrupted Cim classes should not instantiate non-Cim typ
 
         Import-Clixml -Path (Join-Path $PSScriptRoot "assets\CorruptedCim.clixml")
 
+        # Wait up to 10 seconds for calc.exe to run
```
Please use our [`Wait-UntilTrue`](https://github.com/PowerShell/PowerShell/blob/master/test/tools/Modules/HelpersCommon/HelpersCommon.psm1#L1) function.

charlieschmidt replied:
Thats a new function to me, will definitely add that

charlieschmidt replied:
On second reading, may not be appropriate.  WIth other uses of Wait-UntilTrue, we're waiting until something is true, with a 10second time out, checking every second.  Here - we want to wait at least 10 seconds, checking to make sure that calc hasn't started

Maybe just 

```
start-sleep -Seconds 10
Get-Process ... | Should BeNullOrEmpty

?  I'm not as familiar with how a corrupted Cim class could kick off the exe (and have a mac anyway [and the whole thing may not apply in .netcore anyway who knows?]) - I just used the existing Import-Clixml test

iSazonov replied:
You can put `Get-Process Calc` to a script block and send it to `Wait-UntilTrue`

iSazonov replied:
It seems `Microsoft.Management.Infrastructure` is still not ported https://github.com/PowerShell/PowerShell/issues/4562

iSazonov commented on test/powershell/Modules/Microsoft.PowerShell.Utility/clixml.tests.ps1:
```diff
@@ -218,6 +266,22 @@ Describe "Deserializing corrupted Cim classes should not instantiate non-Cim typ
+        $cliXmlContent = Get-Content -Path (Join-Path $PSScriptRoot "assets\CorruptedCim.clixml") -Raw
+        ConvertFrom-CliXml -InputObject $cliXmlContent
+
         # Wait up to 10 seconds for calc.exe to run
```
It is not your code but please use our Wait-UntilTrue function too.

iSazonov reviewed (commented):
@charlieschmidt Thanks for your contribution!

iSazonov commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/XmlCommands.cs:
```diff
@@ -19,95 +19,31 @@
 {
     /// <summary>
-    /// implementation for the Export-Clixml command
+    /// Base implementation for ConvertTo-Clixml and Export-Clixml
```
The comment is public - please add final dot.
Below please review formatting public comments too.

iSazonov commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/XmlCommands.cs:
```diff
@@ -168,45 +155,114 @@ protected override
+        #region FileStream implementation for ToClixmlCommand
+
+        /// <summary>
+        /// handle to file stream
```
Please remove the obvious comment . And why the comment is public for private field? Below too.

iSazonov commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/XmlCommands.cs:
```diff
@@ -351,10 +427,104 @@ protected override void ProcessRecord()
+            if (!successfullyConverted && _inputObjectBuffer.Count > 0)
+            {
+                _helper = new ImportXmlHelper(this, ImportXmlHelper.CreateMemoryStream(string.Join(System.Environment.NewLine, _inputObjectBuffer)));
+                _helper.Import();
```
1. We already create `_helper` - can we reuse it?
2. Can `XmlException` be raised here? If so do we catch it and throw use freadly exception?

adityapatwardhan commented:
@charlieschmidt  Can you respond to some comments from @iSazonov

charlieschmidt commented:
Ponging to say am still following this and will respond with comments or commits tomorrow evening

adityapatwardhan commented:
@charlieschmidt  Please have a look at failing tests. Shall I go ahead and mark this PR as `WIP:` since there is more work expected here.

charlieschmidt commented:
Yes please

adityapatwardhan commented:
@charlieschmidt  Please resolve merge conflict as well. Thanks!
